pyccl/pkresponse.py

```diff
@@ -0,0 +1,795 @@
+
+        T_h[k_use > kmin] = (
+            np.log(pk_hp[k_use > kmin]) - np.log(pk_hm[k_use > kmin])
+        ) / (
+            2 * (np.log(Dp[ia]) - np.log(Dm[ia]))
+        )  # (hp-hm) term is cancelled out
+
+        dpk[k_use <= kmin] = dpklin[k_use <= kmin]
+        dpk[k_use > kmin] = dpknl[k_use > kmin]
+
+        dpk12[ia, :] = pk * (1.0 + (26.0 / 21.0) * T_h - dpk / 3.0)
+
+    if use_log:
+        if np.any(dpk12 <= 0):
+            warnings.warn(
+                "Some values were not positive. "
+                "Will not interpolate in log-space.",
+                category=CCLWarning,
+            )
+            use_log = False
+        else:
+            dpk12 = np.log(dpk12)
+
+    return dpk12
+
+
+def darkemu_Pgm_resp(
+    cosmo,
+    prof_hod,
+    deltah=0.02,
+    log10Mh_min=12.0,
+    log10Mh_max=15.9,
+    lk_arr=None,
+    a_arr=None,
+    extrap_order_lok=1,
+    extrap_order_hik=1,
+    use_log=False,
+):
+    """Implements the response of galaxy-matter power spectrum to
+    the long wavelength modes, described in arXiv:2310.13330.
+    """
+    if lk_arr is None:
+        status = 0
+        nk = lib.get_pk_spline_nk(cosmo.cosmo)
+        lk_arr, status = lib.get_pk_spline_lk(cosmo.cosmo, nk, status)
+        check(status, cosmo=cosmo)
+    if a_arr is None:
+        status = 0
+        na = lib.get_pk_spline_na(cosmo.cosmo)
+        a_arr, status = lib.get_pk_spline_a(cosmo.cosmo, na, status)
+        check(status, cosmo=cosmo)
+
+    k_use = np.exp(lk_arr)
+
+    # Check inputs
+    if not isinstance(prof_hod, halos.profiles.HaloProfile):
+        raise TypeError("prof_hod must be of type `HaloProfile`")
+
+    h = cosmo["h"]
+    k_emu = k_use / h  # [h/Mpc]
+    cosmo.compute_linear_power()
+    pk2dlin = cosmo.get_linear_power("delta_matter:delta_matter")
+
+    # initialize dark emulator class
+    emu = darkemu.de_interface.base_class()
+
+    # set h-modified cosmology to take finite differencing
+    hp = h + deltah
+    hm = h - deltah
+    cosmo_hp, cosmo_hm = set_hmodified_cosmology(cosmo, deltah)
+
+    # Growth factor
+    Dp = cosmo_hp.growth_factor_unnorm(a_arr)
+    Dm = cosmo_hm.growth_factor_unnorm(a_arr)
+
+    na = len(a_arr)
+    nk = len(k_use)
+    dpk12 = np.zeros([na, nk])
+    logMfor_hmf = np.linspace(8, 17, 200)
+    logMh = np.linspace(log10Mh_min, log10Mh_max, 2**5 + 1)  # M_sol/h
+    logM = np.log10(10**logMh / h)
+    Mh = 10**logMh
+    M = 10**logM
+    nM = len(M)
+    dlogM = logM[1] - logM[0]
+    b1_th_tink = np.zeros(nM)
+    Pth = np.zeros((nM, nk))
+    Pnth_hp = np.zeros((nM, nk))
+    Pnth_hm = np.zeros((nM, nk))
+    Pbin = np.zeros((nM, nk))
+
+    nths = np.zeros(nM)
+
+    mass_def = halos.MassDef200m
+    hmf = halos.MassFuncDarkEmulator(mass_def=mass_def)
+    hbf = halos.HaloBiasTinker10(mass_def=mass_def)
+
+    # dark emulator is valid for 0 =< z <= 1.48
+    if np.any(1.0 / a_arr - 1) > 1.5:
+        print("dark emulator is valid for z={:.2f}<1.48")
```

Comment:
hmm, what does `{:.2f}` do here? I think you have to remove that.

Reply:
The `{:.2f}` was unnecessary. I removed that.